### ISSUE TITLE
Vim mode surrounds

### ISSUE THREAD

**Issue by clo4** [zed-industries/zed#4965]
### Check for existing issues

- [X] Completed

### Describe the feature

A very common Vim plugin is [vim-surround](https://github.com/tpope/vim-surround) by the infamous tpope, and the nvim lua rewrite [nvim-surround](https://github.com/kylechui/nvim-surround). These keybinds are muscle memory for me and are the first thing I install when setting up a vimmish editor.

VSCodeVim has these keybinds built in.

I don't think I need to specify any more than this because I can't think of anyone that doesn't use this, but for the sake of keeping the issue self contained:

- `ys[motion][char]` will surround the motion or textobj with char
  - If char is an opening bracket/paren, it will pair it with the close bracket **and insert a space on both sides for padding** eg. with the text `|word`, the command `ysaw{` transforms it to become `{ word }`
  - If char is a closing bracket, the same as above happens but no spaces are inserted.
- `ds[char]` deletes the surrounding characters, or if char is a paired character it deletes the pair.
- `cs[old][new]` changes the surrounding characters from old to new
  - The same rules about opening/closing pairs apply w.r.t inserting a space or not.

The vim-surround readme has a better explanation of all of this of course!

**Comment by mirageN1349** [comment]
This feature is really needed

**Comment by d1y** [comment]
This feature is so important to me, it will take me one step further to mainly use zed❤️

**Comment by weartist** [comment]
If possible, I'd be willing to submit a project to try to provide this functionality

**Comment by clo4** [comment]
Just adding a note for any future implementers/contributors - make sure this works with non-bracket characters too! I sometimes want to surround a word/selection with `\``, `*`, `~`, ` `, and more. I've seen a lot of implementations of surrounds that don't work with ~other characters~

**Comment by weartist** [comment]
> Just adding a note for any future implementers/contributors - make sure this works with non-bracket characters too! I sometimes want to surround a word/selection with `, `*`, `~`, ` `, and more. I've seen a lot of implementations of surrounds that don't work with other characters

Can you give an example of the last two characters, just curious

**Comment by alber70g** [comment]
`viwS"`
* `v` visual 
* `i` inner 
* `w` word 
* `S` surround selection with
* `"` "

Can be used with anything visual so the important part is `S<something>`

Examples:
* `viwS"`: `this is the t<cursor>ext` becomes `this is the "text"`
* `VS"`: 
  ```
  this is a <cursor>sentence
  ```
  becomes
  ```
  "
  this is a sentence
  "
  ```
* `0v$S(`: 
  ```
  this is a <cursor>sentence
  ```
  becomes
  ```
  (this is a sentence)
  ```

**Comment by instantaphex** [comment]
This is definitely a must have for me as well!  Awesome editor, if this feature lands zed will definitely be a daily driver for me.

**Comment by weartist** [comment]
I have written a simplest version of the code for this function. I am not sure whether I should submit the most basic implementation first or wait until it is improved.

**Comment by isiktantanis** [comment]
Literally waiting on this for zed to be my daily driver. Everything else i use and see are great so far, but gotta have surround.

**Comment by weartist** [comment]
I have write a simplified version with only add surround. I will add modification and deletion operations as soon as possible and then provide PR

**Comment by mmtftr** [comment]
+1. I'm struggling with not having this in an otherwise stellar editor.

**Comment by jacksonSingleton** [comment]
+1, only thing holding me back from zed is a fully featured vim-mode

**Comment by romgrk** [comment]
@weartist Still planning on opening a PR? Do you have a partial implementation you can share?

**Comment by weartist** [comment]
> @weartist Still planning on opening a PR? Do you have a partial implementation you can share?

I'm sorry I didn't have much time in the past two weeks, I'll continue this PR next week if I can

**Comment by romgrk** [comment]
Ok, if you don't have enough time or motivation to complete it don't hesitate to open even a partial PR, others can chime in.

**Comment by JosephTLyons** [comment]
Are we ok to close this issue out now that [v0.131.0-pre](https://zed.dev/releases/preview/0.131.0) supports this?

**Comment by 0x2CA** [comment]
@ConradIrwin  doesn't  support `yss`

**Comment by weartist** [comment]
> @ConradIrwin doesn't support `yss`

Yes, the S shortcut is currently occupied, and I'm not sure what to do here, but it shouldn't be too cumbersome to support it in visual mode

**Comment by ConradIrwin** [comment]
I think for `yss` (and `css`) we can start with something like:
```
  {
    "context": "Editor && (vim_operator == ys || vim_operator == cs)",
    "bindings": {
      "s": "vim::CurrentLine"
    }
  },
```

Unfortunately this includes the trailing newline which we don't want to, so will need some code changes too.

**Comment by chenglou** [comment]
Is it too late for a feature request? A shortcut for surrounding the closest pair of things: https://github.com/VSCodeVim/Vim/issues/8699

**Comment by ConradIrwin** [comment]
Not too late! Probably best as a separate issue, or if you'd like to pair on it together: https://calendly.com/conradirwin/pairing

**Comment by bwillem** [comment]
Is there a solution for `S` being occupied? Visual surround - `viwS` doesn't work, because `S` deletes a line and enters insert mode.

**Comment by ConradIrwin** [comment]
Not currently. I think we'd have to build an action for surround in visual mode and then show an example in the docs of how to map `S` to that in visual mode.

If you'd like to work on this with me: https://calendly.com/conradirwin/pairing

**Comment by weartist** [comment]
> I think for `yss` (and `css`) we can start with something like:
> 
> ```
>   {
>     "context": "Editor && (vim_operator == ys || vim_operator == cs)",
>     "bindings": {
>       "s": "vim::CurrentLine"
>     }
>   },
> ```
> 
> Unfortunately this includes the trailing newline which we don't want to, so will need some code changes too.

It seems that we can't bind a series of sequential operations for the same keybinding, for example, if we can, we might be able to configure it directly:
 {
    "context": "Editor && (vim_operator == ys || vim_operator == cs)",
    "bindings": {
      "s": [["vim::Number", 0], "vim::EndOfLine"]
    }
  }

**Comment by quarkw** [comment]
> Is there a solution for `S` being occupied? Visual surround - `viwS` doesn't work, because `S` deletes a line and enters insert mode.

I thought I had commented this a few weeks ago.. but it's pretty easy to simulate this through custom key binds, although you'll have to hardcode every surround character type. And `S` still retains its default functionality, although there's a delay in visual mode as it looks for a keybind chord.

_Note about my comments: I think the vim paste is only needed because I have the copy on yank only setting enabled. If you don't have that enabled, any of these surround commands will pave over your system clipboard. And you would just replace the custom vim paste shortcut with your key bind for normal paste (although vim paste should still work), and you might have to mess with the `left` arrows to make sure everything is in the right spot._

```
  {
    //vim visual mode
    "context": "Editor && vim_mode == visual && !VimWaiting && !VimObject",
    "bindings": {
      "s": "editor::SelectLargerSyntaxNode",
      // shift-s binds (Surround) rely on custom keybinding (ctrl-g p) for vim paste
      "shift-s '": ["workspace::SendKeystrokes", "d i ' ' left left ctrl-g p"],
      "shift-s `": ["workspace::SendKeystrokes", "d i ` ` left left ctrl-g p"],
      "shift-s \"": [
        "workspace::SendKeystrokes",
        "d i \" \" left left ctrl-g p"
      ],
      "shift-s [": [
        "workspace::SendKeystrokes",
        "d i [ space space ] left left left ctrl-g p"
      ],
      "shift-s ]": ["workspace::SendKeystrokes", "d i [ ] left left ctrl-g p"],
      "shift-s {": [
        "workspace::SendKeystrokes",
        "d i { space space } left left left ctrl-g p"
      ],
      "shift-s }": ["workspace::SendKeystrokes", "d i { } left left ctrl-g p"],
      "shift-s (": [
        "workspace::SendKeystrokes",
        "d i ( space space ) left left left ctrl-g p"
      ],
      "shift-s )": ["workspace::SendKeystrokes", "d i ( ) left left ctrl-g p"],
      "[": "editor::SelectLargerSyntaxNode",
      "]": "editor::SelectSmallerSyntaxNode"
    }
  },
  {
    //vim insert mode
    "context": "Editor && vim_mode == insert",
    "bindings": {
      "ctrl-g p": "vim::Paste" //Uses the vim clipboard
    }
  }
```

For anything you don't happen to have hardcoded, you could just do a `cs` command (e.g. `cs'*` to change quotes to asterisks) after doing one of your hardcodes

**Comment by VKondakoff** [comment]
Hmm... I'm just experimenting with surround feature and it looks like there might be a bug with backticks and changing the surround quotes.

Here is an example:

1. `ys[motion][char]`, `ds[char]` and `cs[old][new]` are working great with single and double quotes (`"`, `'`)
1. Now I'm trying to do the same with backtick (\`): 
```
ysiw` - works great (say, word test becomes `test`)
ds` - works great as well (word `test` become test)
cs`" - fails when specifying <old quote> (`)
```

So, `ys` and `ds` are working with backticks well, and `cs`, unfortunately fail.

Can you reproduce this? Is this a bug or am I doing something wrong?

**Comment by ConradIrwin** [comment]
@VKondakoff can reproduce! Looks like the binding for "jump to mark" is taking precedence over the backtick binding...

Will take a look.

cc @oca159 this seems like it's possibly the issue you're hitting in https://github.com/zed-industries/zed/pull/24167 too

**Comment by oca159** [comment]
> @VKondakoff can reproduce! Looks like the binding for "jump to mark" is taking precedence over the backtick binding...
> 
> Will take a look.
> 
> cc @oca159 this seems like it's possibly the issue you're hitting in https://github.com/zed-industries/zed/pull/24167 too

@ConradIrwin it works fine now, the problem was some missing tree sitter queries in brackets.scm, specially for quotes. I added support for multiple popular languages like: js, typescript, rust, go, python, html, css, bash, etc. However, there are still some missing queries for other languages. Please take a look at the pr. It looks good now, as i mentioned the only pending thing is the support for all the languages, however it could take some time for me as i am not an expert in all the languages and i new to write sometimes some custom queries when the nodes for quotes are tricky (take a look at the brackets.scm for go and python)

**Comment by ConradIrwin** [comment]
@VKondakoff currently yes. The plan is to define a "plain text" language that contains various quotes and brackets; but haven't done that yet.

Happy to pair on it if you want to https://cal.com/conradirwin/pairing

**Comment by TKone7** [comment]
> [@VKondakoff](https://github.com/VKondakoff) can reproduce! Looks like the binding for "jump to mark" is taking precedence over the backtick binding...

Is there a workaround for this? I failed to rebind/unbind that default binding.

**Comment by VKondakoff** [comment]
Fixed in 
```
Zed: v0.176.0 (Zed Preview) 
OS: macOS 15.3.1
Memory: 32 GiB
Architecture: x86_64
```

**Comment by quinncomendant** [comment]
> I think we should close this, and (probably) open a follow-up for HTML tag support.

For those of you who read all the comments looking for the answer, I think [this is the issue](https://github.com/zed-industries/zed/issues/10394) that addresses vim-surround HTML tag support.

**Comment by axelson** [comment]
For anyone still coming here, this is what I used to bind `S` to surround:
```
{
  "context": "Editor && vim_mode == visual && !VimWaiting && !VimObject",
  "bindings": {
    "shift-s": "vim::PushAddSurrounds"
  }
}
```
Credit to https://github.com/zed-industries/zed/issues/4965#issuecomment-2562143825, I just fixed a deprecation in the previous syntax.

**Comment by quinncomendant** [comment]
@axelson, why did you use `Editor && vim_mode == visual && !VimWaiting && !VimObject` instead of just `vim_mode == visual`?

The [docs](https://zed.dev/docs/vim#optional-key-bindings) suggest this:

> Vim mode comes with shortcuts to surround the selection in normal mode (ys), but it doesn't have a shortcut to add surrounds in visual mode. By default, shift-s substitutes the selection (erases the text and enters insert mode). To use shift-s to add surrounds in visual mode, you can add the following object to your keymap.
> 
> ```json
> {
>   "context": "vim_mode == visual",
>   "bindings": {
>     "shift-s": "vim::PushAddSurrounds"
>   }
> }
> ```

**Comment by ConradIrwin** [comment]
@quinncomendant Probably from copy-pasting old context values around, they've been deleted for a while now: https://github.com/zed-industries/zed/pull/14233
